### DISCUSSION TITLE
How to type bundles with TypeScript?

### DISCUSSION THREAD

**Discussion by martinbean** [alpinejs/alpine#3444] [1. Help]
Has any one got registering Alpine components from a bundle working nicely with TypeScript? When I try something like this:

```ts
const registerForm = (oldName: string = '', oldEmail: string = '') => ({
    name: oldName,
    email: oldEmail,
});
```
```ts
Alpine.data('registerForm', registerForm);

Alpine.start();
```

I just get the following error:

```
Argument of type '(oldName?: string, oldEmail?: string) => { name: string; email: string; }' is not assignable to parameter of type '(...initialStateArgs: unknown[]) => AlpineComponent<Record<string, any>>'.
  Types of parameters 'oldName' and 'initialStateArgs' are incompatible.
```

How do I get components to play nicely in TypeScript?

**Comment by ekwoka** [comment]
Well, the type definitipns are not part of this package.

But this is mainly because Alpine has sno way for components to be defined in a typesafe way. At some point you'd have to be telling TS what the thing is explicitly

**Reply by martinbean** [reply]
Right? Which is where something like a generic would come in.

Is there really no way to use `Alpine.data` with TypeScript? This seems like a _massive_ oversight.

**Reply by SimoTod** [reply]
Are you using the type definitions for Alpine? The project never targeted typescript but other devs made the effort for adding TS support. For example
https://dev.to/wtho/get-started-with-alpinejs-and-typescript-4dgf

**Reply by ekwoka** [reply]
@martinbean Well, it is an oversight for sure. I agree. But so far Caleb isn't sold on making Alpine in TS (I've actually rewritten it in typescript and am toying with other optimizations along the way).

in many ways it will be difficult to ever make Alpine fully typesafe, since the attribute expressions will need to be just js....well, or would need like an html transformation step...

I will say that the error you are seeing is not one I see at all in active alpine TS projects I have.

But I always define my components directly ino Alpine.data (I pass them to the main to install using Alpine.plugin to keep the definitions in pure ESM modules)

**Reply by martinbean** [reply]
@ekwoka Do you have an example of how you register data components using bundles in TypeScript? As all I do is what I posted above, and then import the module, i.e.

```ts
import Alpine from 'alpinejs';
import registerForm from './alpine/data/register-form';

// Error with below...
Alpine.data('registerForm', registerForm);

Alpine.start();
```

**Reply by ekwoka** [reply]
I do more like this

```ts
// data/registerForm
import type { Alpine } from 'alpinejs'

export const registerForm = (Alpine: Alpine) => 
  Alpine.data('registerForm', (oldName: string = '', oldEmail: string = '') => ({
    name: oldName,
    email: oldEmail,
  })
  
  
// index.ts
import Alpine from 'alpinejs';
import { registerForm } from './data'

Alpine.plugin(registerForm)

Alpine.start()
```

I normally do this with barrel files and such, so I can do something like

```ts
import { data } from './data'
import { stores } from './stores'
import { plugins } from './plugins'

[...plugins, ...stores, ...data].forEach(Alpine.plugin)
```

To keep things a bit more split up and each file small

**Comment by EfrainGuercioy0** [comment]


**Reply by ekwoka** [reply]
Fuck, gpt is making it's way into GitHub discussions....

**Reply by joshhanley** [reply]
@ekwoka should I remove them do you think? 🤔 their answers have nothing to do with @martinbean's question

**Reply by ekwoka** [reply]
I reported them but GitHub mainly just said that the maintainers should block/remove them.

To me they should at least be removed (and our now discussion about it) since its quite obvious it's AI answers that didn't even get the initial post as context.

**Reply by joshhanley** [reply]
@ekwoka ok thanks, good to know. Running it past Caleb, but yeah I agree, I think they should be removed as they're just wrong.

Update: Yeah he agreed to remove, so have done so.
